src/Model/Union.php

```diff
@@ -0,0 +1,464 @@
+
+                break;
+            default:
+                throw (new Exception('Union model does not support this action'))
+                    ->addMoreInfo('mode', $mode);
+        }
+
+        // Substitute FROM table with our subquery expression
+        return parent::action($mode, $args)->reset('table')->table($subquery);
+    }
+
+    /**
+     * Export model.
+     *
+     * @param array|null $fields        Names of fields to export
+     * @param string     $key_field     Optional name of field which value we will use as array key
+     * @param bool       $typecast_data Should we typecast exported data
+     */
+    public function export($fields = null, $key_field = null, $typecast_data = true): array
+    {
+        if ($fields) {
+            $this->onlyFields($fields);
+        }
+
+        $data = [];
+        foreach ($this->getIterator() as $row) {
+            $data[] = $row->get();
+        }
+
+        return $data;
+    }
+
+    /**
+     * Adds nested model in union.
+     *
+     * @param string|Model $class    model
+     * @param array        $fieldMap Array of field mapping
+     */
+    public function addNestedModel($class, array $fieldMap = []): Model
+    {
+        $nestedModel = $this->persistence->add($class);
+
+        $this->union[] = [$nestedModel, $fieldMap];
+
+        return $nestedModel;
+    }
+
+    /**
+     * Specify a single field or array of fields.
+     *
+     * @param string|array $group
+     *
+     * @return $this
+     */
+    public function groupBy($group, array $aggregate = [])
+    {
+        $this->aggregate = $aggregate;
+        $this->group = $group;
+
+        foreach ($aggregate as $fieldName => $seed) {
+            $seed = (array) $seed;
+
+            $field = $this->hasField($fieldName) ? $this->getField($fieldName) : null;
+
+            // first element of seed should be expression itself
+            if (isset($seed[0]) && is_string($seed[0])) {
+                $seed[0] = $this->expr($seed[0], $field ? [$field] : null);
+            }
+
+            if ($field) {
+                $this->removeField($fieldName);
+            }
+
+            $this->addExpression($fieldName, $seed);
+        }
+
+        foreach ($this->union as [$nestedModel, $fieldMap]) {
+            if ($nestedModel instanceof self) {
+                $nestedModel->aggregate = $aggregate;
+                $nestedModel->group = $group;
+            }
+        }
+
+        return $this;
+    }
+
+    /**
+     * Adds condition.
+     *
+     * If Union model has such field, then add condition to it.
+     * Otherwise adds condition to all nested models.
+     *
+     * @param mixed $key
+     * @param mixed $operator
+     * @param mixed $value
+     * @param bool  $forceNested Should we add condition to all nested models?
+     *
+     * @return $this
+     */
+    public function addCondition($key, $operator = null, $value = null, $forceNested = false)
```

Comment:
is union really a SQL UNION? If yes, then we need to apply conditons on all models at query time

getFields() on this Model should return fields on resulting Model

Reply:
There should be logic when you apply conditions on nested models and when on top (_unioned_) model.

Like,
* if field exists in nested model, then apply condition there
* if field is some expression or doesn't exist in nested model, then apply condition on top model itself
Maybe there was something more, but approximately that was idea.

Reply:
to allow good Model nesting/wrapping, we need a native support for it in standard Model